Godeps/_workspace/src/k8s.io/kubernetes/pkg/proxy/userspace/unidlersocket.go

```diff
@@ -0,0 +1,365 @@
+	case "UDP":
+		addr, err := net.ResolveUDPAddr("udp", net.JoinHostPort(host, strconv.Itoa(port)))
+		if err != nil {
+			return nil, err
+		}
+		conn, err := net.ListenUDP("udp", addr)
+		if err != nil {
+			return nil, err
+		}
+		proxySocket := udpProxySocket{UDPConn: conn, port: port}
+		return &udpUnidlerSocket{udpProxySocket: proxySocket, eventRecorder: eventRecorder}, nil
+	}
+	return nil, fmt.Errorf("unknown protocol %q", protocol)
+}
+
+// tcpUnidlerSocket implements proxySocket.  Close() is implemented by net.Listener.  When Close() is called,
+// no new connections are allowed but existing connections are left untouched.
+type tcpUnidlerSocket struct {
+	tcpProxySocket
+	eventRecorder record.EventRecorder
+}
+
+func (tcp *tcpUnidlerSocket) waitForEndpoints(ch chan<- interface{}, service proxy.ServicePortName, proxier *Proxier) {
+	defer close(ch)
+	for {
+		if proxier.loadBalancer.ServiceHasEndpoints(service) {
+			// we have endpoints now, so we're finished
+			return
+		}
+
+		// otherwise, wait a bit before checking for endpoints again
+		time.Sleep(endpointDialTimeout[0])
+	}
+}
+
+func (tcp *tcpUnidlerSocket) acceptConns(ch chan<- net.Conn, myInfo *serviceInfo) {
+	defer close(ch)
+
+	// Block until a connection is made.
+	for {
+		inConn, err := tcp.Accept()
+		if err != nil {
+			if isTooManyFDsError(err) {
+				panic("Accept failed: " + err.Error())
+			}
+
+			// TODO: indicate errors here?
+			if isClosedError(err) {
+				return
+			}
+			if !myInfo.isAlive() {
+				// Then the service port was just closed so the accept failure is to be expected.
+				return
+			}
+			glog.Errorf("Accept failed: %v", err)
+			continue
+		}
+
+		ch <- inConn
+	}
+}
+
+func (tcp *tcpUnidlerSocket) ProxyLoop(service proxy.ServicePortName, myInfo *serviceInfo, proxier *Proxier) {
+	if !myInfo.isAlive() {
+		// The service port was closed or replaced.
+		return
+	}
+
+	// accept connections asynchronously
+	inConns := make(chan net.Conn)
+	go tcp.acceptConns(inConns, myInfo)
+
+	endpointsAvail := make(chan interface{})
+	var allConns *connectionList
+
+	svcName := fmt.Sprintf("%s/%s:%s", service.Namespace, service.Name, service.Port)
+
+TOP_LOOP:
+	for {
+		glog.V(4).Infof("unidling TCP proxy start/reset for service %s/%s:%s", service.Namespace, service.Name, service.Port)
+		// collect connections and wait for endpoints to be available
+		sent_need_pods := false
+		timeout_started := false
+		ticker := time.NewTicker(needPodsTickLen)
+		allConns = newConnectionList(maxHeldConnections, needPodsTickLen, needPodsWaitTimeout, svcName)
+
+	WAIT_LOOP:
+		for {
+			select {
+			case inConn, ok := <-inConns:
+				if !ok {
+					// the listen socket has been closed, so we're finished accepting connections
+					break TOP_LOOP
+				}
+
+				if !sent_need_pods && !proxier.loadBalancer.ServiceHasEndpoints(service) {
+					// TODO: There's a constant in origin for this NeedPods
```

Comment:
Ack.  I've moved most of the stuff into a package in origin, so now I just have a NeedPods constant at the top of this file (which may get moved when I abstract the signal function out anyway).